File: VB/PersonalOrganizer/Common/ViewModel/ReadOnlyCollectionViewModel.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.DataAnnotations
Imports Common.Utils
Imports Common.DataModel
Namespace Common.ViewModel
    Partial Public MustInherit Class ReadOnlyCollectionViewModel(Of TEntity As Class)
        Inherits ReadOnlyCollectionViewModelBase(Of TEntity)
        Public Sub New(Optional ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)) = Nothing)
            MyBase.New(filterExpression)
        End Sub
    End Class
    <POCOViewModel> _
    Public MustInherit Class ReadOnlyCollectionViewModelBase(Of TEntity As Class)
        Private _refreshOnFilterExpressionChanged As Boolean = False
        Private _repository As IReadOnlyRepository(Of TEntity)
        Private _entities As IList(Of TEntity)
        Public Sub New(ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)))
            Me.FilterExpression = filterExpression
            Me._refreshOnFilterExpressionChanged = True
            If Not Me.IsInDesignMode() Then
                OnInitializeInRuntime()
            End If
        End Sub
        Protected Overridable Sub OnInitializeInRuntime()
        End Sub
        Public Overridable Sub Refresh()
            Me._repository = GetRepository()
            Me._entities = GetEntities()
            Me.RaisePropertyChanged(Function(x) Entities)
        End Sub
        Protected ReadOnly Property Repository As IReadOnlyRepository(Of TEntity)
            Get
                If _repository Is Nothing Then
                    _repository = GetRepository()
                End If
                Return _repository
            End Get
        End Property
        Public ReadOnly Property Entities As IList(Of TEntity)
            Get
                If _entities Is Nothing Then
                    _entities = GetEntities()
                End If
                Return _entities
            End Get
        End Property
        Public Overridable Property SelectedEntity As TEntity
        Protected Overridable Sub OnSelectedEntityChanged()
        End Sub
        Public Overridable Property FilterExpression As Expression(Of Func(Of TEntity, Boolean))
        Protected Overridable Sub OnFilterExpressionChanged()
            If _refreshOnFilterExpressionChanged Then
                Refresh()
            End If
        End Sub
        Protected MustOverride Function GetRepository() As IReadOnlyRepository(Of TEntity)
        Protected Overridable Function GetEntities() As IList(Of TEntity)
            Dim queryable = Repository.GetEntities()
            If FilterExpression IsNot Nothing Then
                queryable = queryable.Where(FilterExpression)
            End If
            queryable.Load()
            Return Repository.Local
        End Function
    End Class
End Namespace

File: CS/PersonalOrganizer/Common/ViewModel/ReadOnlyCollectionViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.DataAnnotations;
using System.Collections.ObjectModel;
using System.Threading;
using System.Threading.Tasks;
using PersonalOrganizer.Common.Utils;
using PersonalOrganizer.Common.DataModel;

namespace PersonalOrganizer.Common.ViewModel {
    public partial class ReadOnlyCollectionViewModel<TEntity, TUnitOfWork> : ReadOnlyCollectionViewModel<TEntity, TEntity, TUnitOfWork>
        where TEntity : class
        where TUnitOfWork : IUnitOfWork {

        public static ReadOnlyCollectionViewModel<TEntity, TUnitOfWork> CreateReadOnlyCollectionViewModel(
            IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory,
            Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TEntity>> projection = null) {
            return ViewModelSource.Create(() => new ReadOnlyCollectionViewModel<TEntity, TUnitOfWork>(unitOfWorkFactory, getRepositoryFunc, projection));
        }

        protected ReadOnlyCollectionViewModel(
            IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory,
            Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TEntity>> projection = null)
            : base(unitOfWorkFactory, getRepositoryFunc, projection) {
        }
    }

    /// <summary>
    /// The base class for POCO view models exposing a read-only collection of entities of a given type. 
    /// This is a partial class that provides the extension point to add custom properties, commands and override methods without modifying the auto-generated code.
    /// </summary>
    /// <typeparam name="TEntity">An entity type.</typeparam>
    /// <typeparam name="TUnitOfWork">A unit of work type.</typeparam>
    public partial class ReadOnlyCollectionViewModel<TEntity, TProjection, TUnitOfWork> : ReadOnlyCollectionViewModelBase<TEntity, TProjection, TUnitOfWork>
        where TEntity : class
        where TProjection : class
        where TUnitOfWork : IUnitOfWork {

        public static ReadOnlyCollectionViewModel<TEntity, TProjection, TUnitOfWork> CreateReadOnlyProjectionCollectionViewModel(
            IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory,
            Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TProjection>> projection) {
            return ViewModelSource.Create(() => new ReadOnlyCollectionViewModel<TEntity, TProjection, TUnitOfWork>(unitOfWorkFactory, getRepositoryFunc, projection));
        }

        /// <summary>
        /// Initializes a new instance of the ReadOnlyCollectionViewModel class.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        /// <param name="getRepositoryFunc">A function that returns the repository representing entities of a given type.</param>
        protected ReadOnlyCollectionViewModel(
            IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory,
            Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TProjection>> projection = null)
            : base(unitOfWorkFactory, getRepositoryFunc, projection) {
        }
    }

    /// <summary>
    /// The base class for POCO view models exposing a read-only collection of entities of a given type. 
    /// It is not recommended to inherit directly from this class. Use the ReadOnlyCollectionViewModel class instead.
    /// </summary>
    /// <typeparam name="TEntity">An entity type.</typeparam>
    /// <typeparam name="TUnitOfWork">A unit of work type.</typeparam>
    [POCOViewModel]
    public abstract class ReadOnlyCollectionViewModelBase<TEntity, TProjection, TUnitOfWork> : EntitiesViewModel<TEntity, TProjection, TUnitOfWork>
        where TEntity : class
        where TProjection : class
        where TUnitOfWork : IUnitOfWork {


        /// <summary>
        /// Initializes a new instance of the ReadOnlyCollectionViewModelBase class.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        /// <param name="getRepositoryFunc">A function that returns the repository representing entities of a given type.</param>
        public ReadOnlyCollectionViewModelBase(
            IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory,
            Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc,
            Func<IRepositoryQuery<TEntity>, IQueryable<TProjection>> projection
            ) : base(unitOfWorkFactory, getRepositoryFunc, projection) {
        }


        protected override void OnEntitiesAssigned(Func<TProjection> getSelectedEntityCallback) {
            base.OnEntitiesAssigned(getSelectedEntityCallback);
            SelectedEntity = getSelectedEntityCallback() ?? Entities.FirstOrDefault();
        }

        protected override Func<TProjection> GetSelectedEntityCallback() {
            int selectedItemIndex = Entities.IndexOf(SelectedEntity);
            return () => (selectedItemIndex >= 0 && selectedItemIndex < Entities.Count) ? Entities[selectedItemIndex] : null;
        }

        /// <summary>
        /// The selected enity.
        /// Since ReadOnlyCollectionViewModelBase is a POCO view model, this property will raise INotifyPropertyChanged.PropertyEvent when modified so it can be used as a binding source in views.
        /// </summary>
        public virtual TProjection SelectedEntity { get; set; }

        /// <summary>
        /// The lambda expression used to filter which entities will be loaded locally from the unit of work.
        /// </summary>
        public virtual Expression<Func<TEntity, bool>> FilterExpression { get; set; }

        /// <summary>
        /// Recreates the unit of work and reloads entities.
        /// Since CollectionViewModelBase is a POCO view model, an instance of this class will also expose the RefreshCommand property that can be used as a binding source in views.
        /// </summary>
        public virtual void Refresh() {
            LoadEntities(false);
        }
        public bool CanRefresh() {
            return !IsLoading;
        }

        protected override void OnIsLoadingChanged() {
            base.OnIsLoadingChanged();
            this.RaiseCanExecuteChanged(x => x.Refresh());
        }

        protected virtual void OnSelectedEntityChanged() { }

        protected virtual void OnFilterExpressionChanged() {
            if(IsLoaded || IsLoading)
                LoadEntities(true);
        }

        protected override Expression<Func<TEntity, bool>> GetFilterExpression() {
            return FilterExpression;
        }
    }
}